### PR TITLE
Some Config Validation Fixes

### LINKED ISSUE #2557: Layout fluid invalid value
### Did you clear cache before opening an issue?

- [X] I have cleared my cache

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Does the issue happen when logged in?

Yes

### Does the issue happen when logged out?

Yes

### Does the issue happen in incognito mode when logged in?

Yes

### Does the issue happen in incognito mode when logged out?

Yes

### Account information

{"theme":"8008","themeLight":"serika","themeDark":"serika_dark","autoSwitchTheme":false,"customTheme":true,"customThemeColors":["#000000","#75b8ff","#1100ff","#c2f5ff","#89bef0","#ff4242","#ff0000","#da3333","#791717"],"favThemes":[],"showKeyTips":true,"showLiveWpm":false,"showTimerProgress":true,"smoothCaret":true,"quickTab":true,"punctuation":false,"numbers":false,"words":25,"time":15,"mode":"words","quoteLength":[0],"language":"english","fontSize":"15","freedomMode":false,"difficulty":"master","blindMode":false,"quickEnd":true,"caretStyle":"default","paceCaretStyle":"default","flipTestColors":true,"layout":"qwerty","funbox":"layoutfluid","confidenceMode":"off","indicateTypos":"off","timerStyle":"text","colorfulMode":false,"randomTheme":"off","timerColor":"main","timerOpacity":"0.25","stopOnError":"off","showAllLines":false,"keymapMode":"next","keymapStyle":"staggered","keymapLegendStyle":"lowercase","keymapLayout":"qwerty","fontFamily":"Roboto_Mono","smoothLineScroll":true,"alwaysShowDecimalPlaces":false,"alwaysShowWordsHistory":true,"singleListCommandLine":"manual","capsLockWarning":true,"playSoundOnError":false,"playSoundOnClick":"off","soundVolume":"0.5","startGraphsAtZero":true,"swapEscAndTab":false,"showOutOfFocusWarning":true,"paceCaret":"off","paceCaretCustomSpeed":150,"repeatedPace":true,"pageWidth":"100","chartAccuracy":true,"chartStyle":"scatter","minWpm":"off","minWpmCustomSpeed":100,"highlightMode":"letter","alwaysShowCPM":false,"enableAds":"off","hideExtraLetters":false,"strictSpace":false,"minAcc":"custom","minAccCustom":80,"showLiveAcc":false,"showLiveBurst":false,"monkey":false,"repeatQuotes":"off","oppositeShiftMode":"off","customBackground":"","customBackgroundSize":"cover","customBackgroundFilter":[0,1,1,1,1],"customLayoutfluid":"qwerty#dvorak#colemak","monkeyPowerLevel":"off","minBurst":"off","minBurstCustomSpeed":100,"burstHeatmap":true,"britishEnglish":false,"lazyMode":false,"showAvg":false}

### Current Behavior

When typing in `qwerty dvorak` for `custom layoutfluid`, I am informed that this is an invalid value. This feature has worked for me before in the past, so maybe it's because there needs to be a minimum of 3 layouts?

### Expected Behavior

I'm expecting layoutfluid to allow 2 keyboard layouts

### Steps To Reproduce

1. `esc`
2. `custom layoutfluid`
3. `qwerty dvorak`

### Environment

- OS: Windows 11
- Browser: Firefox
- Browser Version: 97.0.1

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Ferotiq**
### Description

This PR does the following: 
 - allows for between 2 and 5 layouts for custom layoutfluid,
 - added `isConfigValueValidAsync` that returns a promise.
 - moves custom layoutfluid layout validation to the `isConfigValueValidAsync` function.
 - makes an asynchronous `some` function for the `isConfigValueValidAsync` function.
 - removes all `undefined` and `null` checks from the set functions because that's handled by the validator and `undefined` cannot be passed into the set functions (in ts anyway) due to the parameters not having `undefined` as an option. The `apply` function automatically assigns default values for missing properties.
 - Fixes spelling errors.
 - adds boolean to the validator for some set functions that have it in the param type.
 - convert `any` to `unknown`

Closes #2557

### CODE REVIEW

**Comment by Miodec** [comment]
image

Getting conflicting messages on layout fluid save (error and success at the same time).

Also we need a different error for when one of the layouts is incorrect (colema). Currently the error message is misleading.

**Comment by Ferotiq** [comment]
There's a high chance this commit doesn't work yet. I'll have to correct it when I get home.

**Comment by Ferotiq** [comment]
Should all be fixed by this last commit

**Comment by Miodec** [comment]
Still getting an error message, the gray notice disappears too quickly to read fully 
image

**Comment by Miodec** [comment]
Cant open custom theme color picker tab.

**Review by Miodec — CHANGES_REQUESTED** [review]
I think config validation is becoming quite large, lets move it to its own file.

File: frontend/src/scripts/config.ts [hunk 1]
```diff
@@ -277,137 +367,148 @@ export function setMode(mode: MonkeyTypes.Mode, nosave?: boolean): void {
...
+  )
     return invalid("difficulty", diff);
 
-  if (
-    (diff !== "normal" && diff !== "expert" && diff !== "master") ||
-    diff == undefined
-  ) {
+  if (diff !== "normal" && diff !== "expert" && diff !== "master") {
     diff = "normal";
   }
```
Comment (Miodec) [inline frontend/src/scripts/config.ts:429]: These lines are not needed with the cofig validation, correct?

File: frontend/src/scripts/config.ts [hunk 2]
```diff
@@ -277,137 +367,148 @@ export function setMode(mode: MonkeyTypes.Mode, nosave?: boolean): void {
...
+    isConfigValueValid(soe, [["off", "word", "letter"], "boolean"]) ===
+    Validity.INVALID
+  )
     return invalid("stop on error", soe);
 
-  if (soe == undefined || soe === true || soe === false) {
+  if (soe === true || soe === false) {
     soe = "off";
   }
   config.stopOnError = soe;
```
Comment (Miodec) [inline frontend/src/scripts/config.ts:514]: Remove boolean

File: frontend/src/scripts/config.ts [hunk 3]
```diff
@@ -499,192 +601,190 @@ export function setPaceCaret(
...
+  if (
+    isConfigValueValid(val, [["on", "off", "max"], "boolean"]) ===
+    Validity.INVALID
+  )
     return invalid("enable ads", val);
 
-  if (val == undefined || val === true || val === false) {
+  if (val === true || val === false) {
     val = "off";
   }
```
Comment (Miodec) [inline frontend/src/scripts/config.ts:789]: Remove boolean

File: frontend/src/scripts/config.ts [hunk 4]
```diff
@@ -695,88 +795,90 @@ export function setEnableAds(
...
+): boolean {
+  if (
+    isConfigValueValid(val, [["off", "typing"], "boolean"]) === Validity.INVALID
+  )
     return invalid("repeat quotes", val);
 
-  if (val == undefined || val === true || val === false) {
+  if (val === true || val === false) {
     val = "off";
   }
```
Comment (Miodec) [inline frontend/src/scripts/config.ts:813]: Remove boolean

File: frontend/src/scripts/config.ts [hunk 5]
```diff
@@ -1273,119 +1391,145 @@ export function setFreedomMode(freedom: boolean, nosave?: boolean): void {
...
   ConfigEvent.dispatch("customTheme", config.customTheme);
+
+  return true;
 }
 
-export function setTheme(name: string, nosave?: boolean): void {
-  if (!isConfigValueValid(name, ["string"])) return invalid("", name);
+export function setTheme(name: string, nosave?: boolean): boolean {
+  if (isConfigValueValid(name, ["string"]) === Validity.INVALID)
+    return invalid("", name);
```
Comment (Miodec) [inline frontend/src/scripts/config.ts:1461]: Missing first parameter?

File: frontend/src/scripts/config.ts [hunk 6]
```diff
@@ -1273,119 +1391,145 @@ export function setFreedomMode(freedom: boolean, nosave?: boolean): void {
...
 
 function setThemes(
   theme: string,
   customState: boolean,
   nosave?: boolean
-): void {
-  if (!isConfigValueValid(theme, ["string"])) return invalid("", theme);
+): boolean {
+  if (isConfigValueValid(theme, ["string"]) === Validity.INVALID)
+    return invalid("", theme);
```
Comment (Miodec) [inline frontend/src/scripts/config.ts:1499]: Im starting to think this is intended.

File: frontend/src/scripts/config.ts [hunk 7]
```diff
@@ -1273,119 +1391,145 @@ export function setFreedomMode(freedom: boolean, nosave?: boolean): void {
...
+      ["off", "on", "fav", "light", "dark"],
+      "boolean",
+    ]) === Validity.INVALID
   )
     return invalid("random theme", val);
 
-  if (val === undefined || val === true || val === false) {
+  if (val === true || val === false) {
     val = "off";
   }
```
Comment (Miodec) [inline frontend/src/scripts/config.ts:1523]: Remove bool

**Review by Miodec — COMMENTED** [review]
Also, cant we combine `isConfigValueValid` and `invalid` into one function?

File: frontend/src/scripts/config.ts [hunk 8]
```diff
@@ -1158,11 +1264,14 @@ export function setWordCount(
...
+
+  return true;
 }
 
 //caret
-export function setSmoothCaret(mode: boolean, nosave?: boolean): void {
-  if (!isConfigValueValid(mode, ["boolean"])) return invalid("", mode);
+export function setSmoothCaret(mode: boolean, nosave?: boolean): boolean {
+  if (isConfigValueValid(mode, ["boolean"]) === Validity.INVALID)
+    return invalid("", mode);
```
[the remark this answers is not in the document]
Comment (Ferotiq) [reply frontend/src/scripts/config.ts:1274]: lol

File: frontend/src/scripts/config.ts (hunk 2 above)
Comment (Ferotiq) [reply frontend/src/scripts/config.ts:514]: > Also, cant we combine `isConfigValueValid` and `invalid` into one function?

we could

**Pull request merged**